src/pyk/kcfg/exploration.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from pyk.kcfg.kcfg import KCFG
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable, Mapping
+    from typing import Any
+
+    from pyk.kcfg.kcfg import NodeIdLike
+
+
+class KCFGExploration:
+    kcfg: KCFG
+    _terminal: set[int]
+
+    def __init__(self, kcfg: KCFG, terminal: Iterable[NodeIdLike] | None = None) -> None:
+        self.kcfg = kcfg
+        self._terminal = {kcfg._resolve(node_id) for node_id in terminal} if terminal is not None else set()
+
+    #
+    # Recognisers
+    #
+
+    # Terminal node recogniser
+    def is_terminal(self, node_id: NodeIdLike) -> bool:
+        return self.kcfg._resolve(node_id) in self._terminal
+
+    # Explorable node recogniser
+    def is_explorable(self, node_id: NodeIdLike) -> bool:
+        return self.kcfg.is_leaf(node_id) and not self.is_terminal(node_id) and not self.kcfg.is_stuck(node_id)
+
+    #
+    # Collectors
+    #
+
+    # Terminal node collector
+    @property
+    def terminal(self) -> list[KCFG.Node]:
+        return [node for node in self.kcfg.nodes if node.id in self._terminal]
```

Comment:
```suggestion
    # Terminal node collector
    @property
    def terminal(self) -> list[KCFG.Node]:
        return [node for node in self.kcfg.nodes if self.is_terminal(node.id)]
```

This allows us to change th edefinition of `is_terminal` in only one place, and have its effects propogate correctly everywhere.

Reply:
Yes, done